src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            // Remove copyright lines in reverse order.
+            for (var i = removalList.Count - 1; i >= 0; i--)
+            {
+                trivia = trivia.RemoveAt(removalList[i]);
+            }
+
+            var newLineText = document.Project.Solution.Options.GetOption(FormattingOptions.NewLine, root.Language)!;
+            var newLineTrivia = SyntaxGenerator.GetGenerator(document).EndOfLine(newLineText);
+
+            var newHeaderTrivia = CreateNewHeader(leadingSpaces + FileHeaderHelper.CommentPrefix, expectedFileHeader, newLineText);
+
+            // Add a blank line after the header.
+            newHeaderTrivia = newHeaderTrivia.Add(newLineTrivia);
+
+            // Insert header at top of the file.
+            return root.WithLeadingTrivia(newHeaderTrivia.Add(newLineTrivia).AddRange(trivia));
+        }
+
+        private SyntaxNode AddHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            var newLineText = document.Project.Solution.Options.GetOption(FormattingOptions.NewLine, root.Language)!;
+            var newLineTrivia = SyntaxGenerator.GetGenerator(document).EndOfLine(newLineText);
+            var newTrivia = CreateNewHeader(FileHeaderHelper.CommentPrefix, expectedFileHeader, newLineText).Add(newLineTrivia).Add(newLineTrivia);
+
+            // Skip blank lines already at the beginning of the document, since we add our own
+            var leadingTrivia = root.GetLeadingTrivia();
+            var skipCount = 0;
+            for (var i = 0; i < leadingTrivia.Count; i++)
+            {
+                if (leadingTrivia[i].RawKind == SyntaxKinds.EndOfLineTrivia)
+                {
+                    skipCount = i + 1;
+                }
+                else if (leadingTrivia[i].RawKind != SyntaxKinds.WhitespaceTrivia)
+                {
+                    break;
+                }
+            }
+
+            newTrivia = newTrivia.AddRange(leadingTrivia.Skip(skipCount));
+
+            return root.WithLeadingTrivia(newTrivia);
+        }
+
+        private SyntaxTriviaList CreateNewHeader(string prefixWithLeadingSpaces, string expectedFileHeader, string newLineText)
+        {
+            var copyrightText = GetCopyrightText(prefixWithLeadingSpaces, expectedFileHeader, newLineText);
+            var newHeader = copyrightText;
+            return ParseLeadingTrivia(newHeader);
+        }
+
+        private static string GetCopyrightText(string prefixWithLeadingSpaces, string copyrightText, string newLineText)
+        {
+            copyrightText = copyrightText.Replace("\r\n", "\n");
+            var lines = copyrightText.Split('\n');
+            return string.Join(newLineText, lines.Select(line =>
+            {
+                if (string.IsNullOrEmpty(line))
+                {
+                    return prefixWithLeadingSpaces;
```

Comment:
can you doc this case?  i'm not quite getting it.

Reply:
Avoids putting trailing whitespace on blank lines, for a file header that has blank lines as part of the header itself. Covered by `TestHeaderMissingRequiredNewLinesAsync` but will add a comment here.